### PR TITLE
refactor: subnets table view

### PR DESCRIPTION
**Pull request by petermakowski**
## Done

- refactor: subnets table view

### Note:
- additional improvements will be investigated as part of https://github.com/canonical-web-and-design/app-tribe/issues/681
  - data model and implementation for table rows groups could use some work, followed the structure in legacy implementation for now (adding `isVisuallyHidden` key to each subsequent item in the group)
  - I made some work into UI performance  (only a single row re-renders on update, instead of the entire table) but this needs to be investigated further - we need to make sure this works with a very large dataset (e.g. few thousands of entries)
  - implementation could be refactored to be closer to how machines list (e.g. search pattern in `machine/utils/search.ts`)

### No space warning toggle
image

image
_Using a button toggle instead of tooltip for better accessibility_

### Group by more information toggle
image
_Collapsed_

image
_Expanded_

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to `MAAS/r/networks?by=fabric` and try breaking the table by adding different items / changing the view from/to group by fabric/space

## Fixes

Fixes: # .

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against *master* rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in master.

### CODE REVIEW

File: ui/src/app/subnets/views/SubnetsList/SubnetsTable/SubnetsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  TableHeader,
+  FabricRow,
+  SpaceRow,
+} from "app/subnets/views/SubnetsList/SubnetsTable/components";
+import type {
+  SubnetsTableRow,
+  SubnetGroupByProps,
+} from "app/subnets/views/SubnetsList/SubnetsTable/types";
+
+const FabricTable = ({
```
Comment (Caleb-Ellis) [inline ui/src/app/subnets/views/SubnetsList/SubnetsTable/SubnetsTable.tsx:16]: I think each of the Fabric and Space-specific components could live in separate subdirectories. There's a bit of cognitive overhead in trying to figure out what's relevant and what's not in these component files at the moment.

File: ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+);
+
+export const FabricRow = memo(
+  ({ columns }: SubnetsTableRow): JSX.Element => (
+    <TableRowComponent
+      aria-label={columns.fabric.label || undefined}
+      style={{
+        transition: "opacity 250ms",
+        opacity: 1,
+      }}
```
Comment (Caleb-Ellis) [inline ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx:57]: What does this do? Could it be a class instead?

File: ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+export const SpaceRow = memo(({ columns }: SubnetsTableRow): JSX.Element => {
+  const [isWarningOpen, setIsWarningOpen] = useState(false);
+  return (
+    <TableRowComponent
+      aria-label={columns.fabric.label || undefined}
+      style={{
+        transition: "opacity 250ms",
+        opacity: 1,
+      }}
```
Comment (Caleb-Ellis) [inline ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx:120]: As above

**Review by Caleb-Ellis — APPROVED** [review]
This looks great! I like the light-touch improvements to accessibility.

So I know I said in standup that big PRs would be easier due to the timezone difference, but this was actually quite difficult to review. Not so much the lines of code, but more the number of different things to address e.g. fabric table, space table, group by select, general component structure, general folder structure, etc. In the future lets try and do multiple smaller PRs instead.

Anyway I think this is probably fine to merge, but it'd be good if we could have a quick call about some of the decisions you made first.

File: ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx (hunk 2 above)
Comment (petermakowski) [reply ui/src/app/subnets/views/SubnetsList/SubnetsTable/components.tsx:57]: Sorry, this shouldn't be here (this is back from when I considered transitioning in new rows). One more reason for smaller PRs, missed it.

File: ui/src/app/store/subnet/utils.ts [hunk 4]
```diff
@@ -26,3 +28,32 @@ export const getSubnetDisplay = (
...
+ */
+export const getAvailableIPs = (subnet: Subnet | null | undefined): string => {
+  if (!subnet) {
+    return "Unconfigured";
+  } else {
+    return `${subnet?.statistics?.available_string}`;
+  }
+};
+
+export const getSubnetsInVLAN = (
```
[the remark this answers is not in the document]
Comment (petermakowski) [reply ui/src/app/store/subnet/utils.ts:45]: Personally I don't find them very useful (especially considering we're using typescript), but will add for consistency.

File: integration/cypress/integration/utils.ts [hunk 5]
```diff
@@ -14,3 +14,4 @@ export const generateMac = () =>
 
 export const generateEmail = () => `${nanoid()}@example.com`;
 export const generateId = () => nanoid();
+export const generateVid = () => `${Math.floor(Math.random() * 1000)}`;
```
[the remark this answers is not in the document]
Comment (petermakowski) [reply integration/cypress/integration/utils.ts:17]: Fair point, and I did think about it - but decided it's good enough for now.

**Pull request merged**